### PR TITLE
Add background callback + more drawWsg optimizations

### PR DESCRIPTION
**Pull request by cnlohr**

### CODE REVIEW

File: main/display/display.h [hunk 1]
```diff
@@ -34,13 +34,15 @@ typedef struct
...
 } wsg_t;
 
+struct display;
+
 typedef void (*pxSetFunc_t)(int16_t x, int16_t y, paletteColor_t px);
 typedef paletteColor_t (*pxGetFunc_t)(int16_t x, int16_t y);
 typedef paletteColor_t* (*pxFbGetFunc_t)(void);
 typedef void (*pxClearFunc_t)(void);
-typedef void (*drawDisplayFunc_t)(bool drawDiff);
+typedef void (*drawDisplayFunc_t)(bool drawDiff,void (*fnBackgroundDrawCallback)(struct display* disp, int16_t x, int16_t y, int16_t w, int16_t h, int16_t up, int16_t upNum ));
```
Comment (AEFeinstein) [inline main/display/display.h:43]: Please typedef the new function pointer. It's used plenty of places.

File: components/hdw-oled/ssd1306.c [hunk 2]
```diff
@@ -379,6 +379,8 @@ uint8_t priorFb[(OLED_WIDTH * (OLED_HEIGHT / 8))] = {0};
 
 bool fbChanges = false;
 
+display_t oledDisp;
```
Comment (AEFeinstein) [inline components/hdw-oled/ssd1306.c:382]: A part of me wanted to have these functions be stateless and have everything in `display_t`, but that ended up being a little too idealistic for this project.

Comment (AEFeinstein) [reply components/hdw-oled/ssd1306.c:382]: Also this should be a pointer.

File: main/display/display.h [hunk 3]
```diff
@@ -49,7 +51,9 @@ typedef struct
     uint16_t w;
     uint16_t h;
     paletteColor_t * pxFb; // may be null
-} display_t;
+};
+
+typedef struct display display_t;
```
Comment (AEFeinstein) [inline main/display/display.h:56]: Any particular reason for this shuffle?

**Review by AEFeinstein — CHANGES_REQUESTED** [review]

File: components/hdw-oled/ssd1306.c (hunk 2 above)
Comment (cnlohr) [reply components/hdw-oled/ssd1306.c:382]: your wish is my command

File: main/display/display.h (hunk 3 above)
Comment (cnlohr) [reply main/display/display.h:56]: Yes, it's the only way I could get the definitions to work.

File: main/display/display.h (hunk 1 above)
Comment (cnlohr) [reply main/display/display.h:43]: done

File: main/swadge_esp32.c [hunk 4]
```diff
@@ -587,10 +587,11 @@ void mainSwadgeTask(void* arg __attribute((unused)))
                 }
 
                 // Draw the display at the given frame rate
+
 #ifdef OLED_ENABLED
-                oledDisp.drawDisplay(true);
+                oledDisp.drawDisplay(&tftDisp, true, cSwadgeMode->fnBackgroundDrawCallback);
```
Comment (AEFeinstein) [inline main/swadge_esp32.c:592]: Typo

**Review by AEFeinstein — CHANGES_REQUESTED** [review]

**Comment by AEFeinstein** [comment]
I'll allow the reboot thing here. Still got that typo though!

**Review by AEFeinstein — APPROVED** [review]

**Pull request merged**